src/proxy.rs

```diff
@@ -151,69 +147,63 @@ impl Proxy {
         let metrics = Arc::new(metrics);
         let socket_factory = Arc::new(DefaultSocketFactory);
 
-        let pi = ProxyInputs {
+        let pi = ProxyInputs::new(
             cfg,
-            state,
             cert_manager,
-            connection_manager: ConnectionManager::default(),
+            ConnectionManager::default(),
+            state,
             metrics,
-            hbone_port: 0,
             socket_factory,
-            proxy_workload_info: None,
-        };
+            None,
+        );
         Self::from_inputs(pi, drain).await
     }
-    pub(super) async fn from_inputs(mut pi: ProxyInputs, drain: Watch) -> Result<Self, Error> {
-        // illegal_ports are internal ports that clients are not authorized to send to
-        let mut illegal_ports: HashSet<u16> = HashSet::new();
+    pub(super) async fn from_inputs(mut pi: Arc<ProxyInputs>, drain: Watch) -> Result<Self, Error> {
         // We setup all the listeners first so we can capture any errors that should block startup
         let inbound = Inbound::new(pi.clone(), drain.clone()).await?;
-        pi.hbone_port = inbound.address().port();
-        //  HBONE doesn't have redirection, so we cannot have loops, but this would allow multiple layers of HBONE.
-        // This might be desirable in the future, but for now just ban it.
-        illegal_ports.insert(inbound.address().port());
+
+        // This exists for `direct` integ tests, no other reason
+        if pi.cfg.fake_self_inbound {
```

Comment:
nit: could `#[cfg(any(test, feature = "testing"))]` it. Do we want to use tracing instead of println?

Reply:
> could #[cfg(any(test, feature = "testing"))]

Yeah I considered that. In this case it shouldn't hurt because it's already a config-based opt-in just for tests.

Reply:
Ah thats fair. I was thinking this was like fake_ca where we do it to avoid someone doing something dumb in prod -- but that one is configurable by env var; this is not, so it should be impossible

Reply:
I'm adding it anyway tho, might as well save the check in real code.